File: backend-spring-boot/src/main/java/no/vicx/backend/esport/EsportService.java
package no.vicx.backend.esport;

import no.vicx.backend.esport.vm.EsportVm;
import no.vicx.backend.esport.vm.MatchType;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
public class EsportService {

    private final EsportClient esportClient;

    public EsportService(EsportClient esportClient) {
        this.esportClient = esportClient;
    }

    @Cacheable(value = "ESPORT")
    public Mono<EsportVm> getMatches() {
        return Mono.zip(
                esportClient.getMatches(MatchType.running).collectList(),
                esportClient.getMatches(MatchType.upcoming).collectList()
        ).map(tuple -> new EsportVm(
                tuple.getT1(),
                tuple.getT2()));
    }
}

File: backend-spring-boot/src/main/kotlin/no/vicx/backend/esport/EsportService.kt
package no.vicx.backend.esport

import no.vicx.backend.esport.vm.EsportVm
import no.vicx.backend.esport.vm.MatchType
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class EsportService(
    private val esportClient: EsportClient,
) {
    @Cacheable(value = ["ESPORT"])
    fun getMatches(): Mono<EsportVm> =
        Mono
            .zip(
                esportClient.getMatches(MatchType.RUNNING).collectList(),
                esportClient.getMatches(MatchType.UPCOMING).collectList(),
            ).map { tuple -> EsportVm(tuple.t1, tuple.t2) }
}
